x-pack/plugins/apm/public/components/shared/Links/url_helpers.ts

```diff
@@ -6,95 +6,89 @@
-}
-
-function stringifyWithoutEncoding(query: QueryParams) {
-  return qs.stringify(query, undefined, undefined, {
-    encodeURIComponent: (v: string) => v
-  });
-}
+function createG(values: APMQueryParams) {
+  const g: RisonDecoded['_g'] = {};
 
-function risonSafeDecode(value?: string) {
-  if (!value) {
-    return {};
+  if (typeof values.rangeFrom !== 'undefined') {
+    set(g, 'time.from', encodeURIComponent(values.rangeFrom));
+  }
+  if (typeof values.rangeTo !== 'undefined') {
+    set(g, 'time.to', encodeURIComponent(values.rangeTo));
   }
 
-  try {
-    const decoded = rison.decode(value);
-    return isPlainObject(decoded) ? (decoded as StringMap) : {};
-  } catch (e) {
-    return {};
+  if (typeof values.refreshPaused !== 'undefined') {
+    set(g, 'refreshInterval.pause', String(values.refreshPaused));
+  }
+  if (typeof values.refreshInterval !== 'undefined') {
+    set(g, 'refreshInterval.value', String(values.refreshInterval));
   }
+
+  return g;
 }
 
-// Kibana default set in: https://github.com/elastic/kibana/blob/e13e47fc4eb6112f2a5401408e9f765eae90f55d/x-pack/plugins/apm/public/utils/timepicker/index.js#L31-L35
-// TODO: store this in config or a shared constant?
-const DEFAULT_KIBANA_TIME_RANGE = {
-  time: {
-    from: 'now-24h',
-    mode: 'quick',
-    to: 'now'
-  }
-};
+export const PERSISTENT_APM_PARAMS = [
+  'kuery',
+  'rangeFrom',
+  'rangeTo',
+  'refreshPaused',
+  'refreshInterval'
+];
 
-function getQueryWithRisonParams(
+function getSearchString(
   location: Location,
   pathname: string,
-  query: RisonDecoded = {}
+  { _g, _a, ...query }: APMQueryParams & RisonDecoded = {}
 ) {
-  // Preserve current _g and _a
   const currentQuery = toQuery(location.search);
-  const decodedG = risonSafeDecode(currentQuery._g);
-  const combinedG = { ...DEFAULT_KIBANA_TIME_RANGE, ...decodedG, ...query._g };
-  const encodedG = rison.encode(combinedG);
-  const encodedA = query._a ? rison.encode(query._a) : '';
-
-  const nextQuery: StringMap = {
-    ...query,
-    _g: encodedG
+  const nextQuery = {
+    ...pick(currentQuery, PERSISTENT_APM_PARAMS),
+    ...query
   };
 
-  // Preserve kuery for apm links
+  // Preserve existing params for apm links
   const isApmLink = pathname.includes('app/apm') || pathname === '';
-  if (currentQuery.kuery && isApmLink) {
-    nextQuery.kuery = currentQuery.kuery;
+  if (isApmLink) {
+    return fromQuery(nextQuery);
   }
 
+  // Create _g value for non-apm links
+  const g = createG(nextQuery);
+  const encodedG = rison.encode(g);
+  const encodedA = _a ? rison.encode(_a) : ''; // TODO: Do we need to url-encode the _a values before rison encoding _a?
+  const risonQuery: RisonEncoded = {
+    _g: encodedG
+  };
+
   if (encodedA) {
-    nextQuery._a = encodedA;
+    risonQuery._a = encodedA;
   }
 
-  return nextQuery;
+  // don't URI-encode the already-encoded rison
+  return qs.stringify(risonQuery, undefined, undefined, {
+    encodeURIComponent: (v: string) => v
+  });
 }
```

Comment:
It would be really great if we can get Rison out of url_helpers completely, and make it a separate converter, that `DiscoverLink` and `MLJobLink` can use. This way it's separated from how urls are handled internally in APM.

Reply:
Mm, this is a good idea. I'll take a stab at it and ping you if I have trouble figuring out how to make this work cleanly.